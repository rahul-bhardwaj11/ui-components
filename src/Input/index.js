import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AntInput from 'antd/lib/input';
import 'antd/lib/input/style/index.css';
import styled from 'styled-components';
import theme from '../styles/theme';
import mixins from '../styles/mixins';
import classnames from 'classnames';

const noop = () => undefined;

const isString = value => {
  return typeof value === 'string';
};

const MtInput = styled.div`
  position: relative;
  .counterStyle {
    ${mixins.smallGreyLink()};
    line-height: 20px;
    position: absolute;
    right: 0px;
  }
  .displayN {
    display: none;
  }
  .error {
    font-size: 12px;
    bottom: -20px;
    left: 2px;
    position: absolute;
    color: ${theme.colors.BITTERSWEET};
  }
  .ant-input.errorInputStyle {
    border: 1px solid red;
    &:focus,
    &:hover {
      border: 1px solid #f5222d;
    }
  }

  .ant-input {
    border: 1px solid ${theme.colors.ALTO};
    font-family: inherit;
  }
  .ant-input {
    &:hover {
      border-color: ${theme.colors.SILVER};
    }
  }
  .ant-input {
    &.active,
    &:focus {
      border-color: ${theme.colors.SILVER};
      outline: 0;
      box-shadow: none;
    }
  }
  .ant-input-disabled {
    border-color: ${theme.colors.DISABLE};
    background-color: #f8f8f8;
    color: ${theme.colors.SILVER};
    &:hover {
      border-color: ${theme.colors.DISABLE};
    }
  }
  .ant-input-error {
    border-color: #ff6060;
  }
`;

class Input extends Component {
  static propTypes = {
    className: PropTypes.string,
    placeholder: PropTypes.string,
    maxLength: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
    maxLengthClassName: PropTypes.string,
    wrapperClassName: PropTypes.string,
    showMaxLength: PropTypes.bool,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    defaultValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    type: PropTypes.oneOf(['text', 'number', 'password', 'file']),
    errors: PropTypes.array
  };

  static defaultProps = {
    onChange: noop,
    onFocus: noop,
    errors: [],
    type: 'text'
  };

  constructor(props) {
    super(props);
    this.state = {
      value: this.handleValue(this.props.value)
    };
  }

  handleValue = value => {
    let trimmedValue = isString(value) ? value.trim() : value;
    trimmedValue = trimmedValue || '';
    value = trimmedValue.length ? value : trimmedValue;
    const { maxLength } = this.props;
    if (maxLength) {
      return maxLength <= value.length ? value : value.substring(0, maxLength);
    }
    return value;
  };

  onChange = event => {
    const { onChange } = this.props;
    const value = this.handleValue(event.target.value);
    this.setState({ value });
    onChange(event, value);
  };

  moveCaretAtEnd = e => {
    var temp_value = e.target.value;
    e.target.value = '';
    e.target.value = temp_value;
    this.props.onFocus(e);
  };

  componentWillMount() {
    let { value, defaultValue } = this.props;
    if (!value) {
      value = defaultValue ? defaultValue : '';
    }

    this.setState({ value: this.handleValue(value) });
  }

  componentWillReceiveProps(newProps) {
    const { value } = this.props;
    let { value: newValue } = newProps;
    if (newValue !== value) {
      this.setState({ value: this.handleValue(newValue) });
    }
  }

  render() {
    const {
      errors,
      maxLength,
      maxLengthClassName,
      showMaxLength,
      wrapperClassName,
      className,
      ...rest
    } = this.props;
    const { value } = this.state;
    return (
      <MtInput className={wrapperClassName}>
        <AntInput
          {...rest}
          value={value}
          maxLength={maxLength}
          onChange={this.onChange}
          onFocus={this.moveCaretAtEnd}
          className={classnames(className, { errorInputStyle: errors[0] })}
        />
        {maxLength &&
          showMaxLength && (
            <div
              key="maxLength"
              className={classnames(
                { ['counterStyle']: maxLength, displayN: !maxLength },
                maxLengthClassName
              )}
            >
              {maxLength && maxLength - value.length}
            </div>
          )}
        <div
          key="error"
          className={classnames(errors[0] ? 'error' : 'displayN')}
        >
          {errors[0]}
        </div>
      </MtInput>
    );
  }
}

export default Input;
