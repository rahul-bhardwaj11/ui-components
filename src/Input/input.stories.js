import React from 'react';

import { storiesOf } from '@storybook/react';
import Input from './index';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, text } from '@storybook/addon-knobs';

const stories = storiesOf('Input', module);
stories.addDecorator(withKnobs);

const styles = {
  width: '200px'
};

stories
  .add(
    'Default Input',
    withInfo('Basic usage of the Input')(() => (
      <div style={styles}>
        <Input placeholder={text('Placeholder', 'Placeholder')} />
      </div>
    ))
  )
  .add(
    'Input with Numeric value',
    withInfo('Basic usage of the Input')(() => (
      <div style={styles}>
        <Input
          placeholder={text('Placeholder', 'Placeholder')}
          type="number"
          max={5}
        />
      </div>
    ))
  )
  .add(
    'Input with maxChar length',
    withInfo('Basic usage of the Input')(() => (
      <div style={styles}>
        <Input
          placeholder={text('Placeholder', 'Placeholder')}
          showMaxLength={true}
          maxLength={40}
        />
      </div>
    ))
  );
