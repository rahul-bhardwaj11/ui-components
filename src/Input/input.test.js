import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Input from './index';

describe('Input', () => {
  it('matches default snapshot', () => {
    const tree = renderer
      .create(<Input placeholder={'Placeholder'} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should handle click', () => {
    const stopPropagation = jest.fn();
    const input = shallow(<Input placeholder={null} />);
    input.find('Input').simulate('click', { stopPropagation });
    expect(stopPropagation).toHaveBeenCalled();
  });
});
