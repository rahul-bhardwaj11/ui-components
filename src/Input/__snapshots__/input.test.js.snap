// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches default snapshot 1`] = `
<div
  className="sc-bdVaJa gEilJX"
>
  <input
    className="ant-input"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    placeholder="Placeholder"
    type="text"
  />
</div>
`;
